src/sync.rs

```diff
@@ -28,10 +71,98 @@ pub fn sync_local_to_remote(local_dir_absolute_path: &Path, config: &Config, ign
         project_dir_on_remote_machine = project_dir_on_remote_machine(local_dir_absolute_path))
     );
 
-    execute_rsync(&mut command)
+    match execute_rsync(&mut command) {
+        Err(reason) => Err(PushErr {
+            duration: start_time.elapsed(),
+            message: reason,
+        }),
+        Ok(_) => Ok(PushOk {
+            duration: start_time.elapsed()
+        }),
+    }
+}
+
+pub fn pull(local_dir_absolute_path: &Path, config: Config, ignore: Ignore, pull_mode: &PullMode, remote_command_finished_signal: BusReader<Result<RemoteCommandOk, RemoteCommandErr>>) -> Receiver<Result<PullOk, PullErr>> {
+    match pull_mode {
+        PullMode::Serial => pull_serial(local_dir_absolute_path.to_path_buf(), config, ignore, remote_command_finished_signal),
+        PullMode::Parallel(pause_between_pulls) => pull_parallel(local_dir_absolute_path.to_path_buf(), config, ignore, *pause_between_pulls, remote_command_finished_signal)
+    }
+}
+
+fn pull_serial(local_dir_absolute_path: PathBuf, config: Config, ignore: Ignore, mut remote_command_finished_rx: BusReader<Result<RemoteCommandOk, RemoteCommandErr>>) -> Receiver<Result<PullOk, PullErr>> {
+    let (pull_finished_tx, pull_finished_rx): (Sender<Result<PullOk, PullErr>>, Receiver<Result<PullOk, PullErr>>) = unbounded();
+
+    #[allow(unused_must_use)] // We don't handle remote_command_result, in any case we need to pull after it.
+    thread::spawn(move || {
+        remote_command_finished_rx
+            .recv()
+            .expect("Could not receive remote_command_finished_rx");
+
+        pull_finished_tx
+            .send(_pull(local_dir_absolute_path.as_path(), &config, &ignore))
+            .expect("Could not send pull_finished signal");
+    });
+
+    pull_finished_rx
+}
+
+fn pull_parallel(local_dir_absolute_path: PathBuf, config: Config, ignore: Ignore, pause_between_pulls: Duration, mut remote_command_finished_signal: BusReader<Result<RemoteCommandOk, RemoteCommandErr>>) -> Receiver<Result<PullOk, PullErr>> {
+    let (pull_finished_tx, pull_finished_rx): (Sender<Result<PullOk, PullErr>>, Receiver<Result<PullOk, PullErr>>) = unbounded();
+    let start_time = Instant::now();
+
+    thread::spawn(move || {
+        let mut should_run = true;
+
+        while should_run {
+            match _pull(local_dir_absolute_path.as_path(), &config, &ignore) {
+                Err(err) => {
+                    should_run = false;
+                    pull_finished_tx
+                        .send(Err(err)) // TODO handle code 24.
+                        .expect("Could not send pull_finished signal");
+                }
+                Ok(_) => thread::sleep(pause_between_pulls),
+            }
+
+            match remote_command_finished_signal.try_recv() {
+                Err(reason) => match reason {
+                    Disconnected => should_run = false,
+                    Empty => thread::sleep(pause_between_pulls)
```

Comment:
Seems like most iterations would actually sleep for 2x `pause_between_pulls`: once above when `_pull` succeeds normally, and once here while the remote command is still executing.

Reply:
Good catch! I actually only need to sleep while remote command is execution, I think I had that before but then broke it during refactoring, thanks